libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2549,61 +2571,66 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+				auto const& t = funType->returnParameterTypes();
+
+				if (memberName == "value")
+				{
+					if (funType->kind() == FunctionType::Kind::Creation)
+						return { 8827_error, "Constructor for " + t.front()->toString() + " must be payable for member \"value\" to be available." };
+					else if (
+						funType->kind() == FunctionType::Kind::DelegateCall ||
+						funType->kind() == FunctionType::Kind::BareDelegateCall
+					)
+						return { 8477_error, "Member \"value\" is not allowed in delegated calls due to \"msg.value\" persisting." };
+					else
+						return { 8820_error, "Member \"value\" is only available for payable functions." };
+				}
 				else if (
-					funType->kind() == FunctionType::Kind::DelegateCall ||
-					funType->kind() == FunctionType::Kind::BareDelegateCall
+					t.size() == 1 &&
+					(t.front()->category() == Type::Category::Struct ||
+					t.front()->category() == Type::Category::Contract)
 				)
-					errorMsg = "Member \"value\" is not allowed in delegated calls due to \"msg.value\" persisting.";
-				else
-					errorMsg = "Member \"value\" is only available for payable functions.";
+					return { 6005_error, errorMsg + " Did you intend to call the function?" };
 			}
-			else if (
-				t.size() == 1 &&
-				(t.front()->category() == Type::Category::Struct ||
-				t.front()->category() == Type::Category::Contract)
-			)
-				errorMsg += " Did you intend to call the function?";
-		}
-		else if (exprType->category() == Type::Category::Contract)
-		{
-			for (auto const& addressMember: TypeProvider::payableAddress()->nativeMembers(nullptr))
-				if (addressMember.name == memberName)
-				{
-					Identifier const* var = dynamic_cast<Identifier const*>(&_memberAccess.expression());
-					string varName = var ? var->name() : "...";
-					errorMsg += " Use \"address(" + varName + ")." + memberName + "\" to access this address member.";
-					break;
-				}
-		}
-		else if (auto addressType = dynamic_cast<AddressType const*>(exprType))
-		{
-			// Trigger error when using send or transfer with a non-payable fallback function.
-			if (memberName == "send" || memberName == "transfer")
+			else if (exprType->category() == Type::Category::Contract)
 			{
-				solAssert(
-					addressType->stateMutability() != StateMutability::Payable,
-					"Expected address not-payable as members were not found"
-				);
+				for (auto const& addressMember: TypeProvider::payableAddress()->nativeMembers(nullptr))
+					if (addressMember.name == memberName)
+					{
+						Identifier const* var = dynamic_cast<Identifier const*>(&_memberAccess.expression());
+						string varName = var ? var->name() : "...";
+						errorMsg += " Use \"address(" + varName + ")." + memberName + "\" to access this address member.";
+						return { 5256_error, errorMsg };
+					}
+			}
+			else if (auto addressType = dynamic_cast<AddressType const*>(exprType))
```

Comment:
To be more consistent with the rest of PR where you extract `const` and `*/&` from `auto`
```suggestion
			else if (auto const* addressType = dynamic_cast<AddressType const*>(exprType))
```